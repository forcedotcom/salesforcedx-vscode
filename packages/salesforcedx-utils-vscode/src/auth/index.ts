/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

/**
 * This file is intentionally left blank. AuthUtil is exported
 * in the root of salesforcedx-utils-vscode.
 */
