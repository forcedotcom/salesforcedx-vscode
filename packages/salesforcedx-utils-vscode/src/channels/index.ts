/*
 * Copyright (c) 2017, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

import { ChannelService } from './channelService';
export const channelService = ChannelService.getInstance();
export { DEFAULT_SFDX_CHANNEL } from './channelService';
export { ChannelService } from './channelService';
