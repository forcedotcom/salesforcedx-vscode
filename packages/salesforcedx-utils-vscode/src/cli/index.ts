/*
 * Copyright (c) 2017, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

export interface Command {
  readonly command: string;
  readonly description?: string;
  readonly args: string[];
  readonly logName?: string;

  toString(): string;
  toCommand(): string;
}

export { CommandBuilder, SfdxCommandBuilder } from './commandBuilder';
export {
  CliCommandExecutor,
  CliCommandExecution,
  CommandExecution,
  CompositeCliCommandExecutor,
  GlobalCliEnvironment
} from './commandExecutor';
export { CommandOutput } from './commandOutput';
export { ForceConfigGet } from './forceConfigGet';
export { ForceOrgDisplay, OrgInfo } from './forceOrgDisplay';
export { LocalCommandExecution } from './localCommandExecutor';
export {
  ForceDeployResultParser,
  ForceSourceDeployErrorResponse,
  ForceSourceDeploySuccessResponse,
  DeployResult,
  CONFLICT_ERROR_NAME
} from './deployResultParser';
export { TestRunner } from './testRunner';
export {
  OrgCreateSuccessResult,
  OrgCreateErrorResult,
  OrgCreateResultParser
} from './orgCreateResultParser';
export {
  OrgOpenSuccessResult,
  OrgOpenContainerResultParser,
  OrgOpenErrorResult
} from './orgOpenContainerResultParser';
export {
  DiffErrorResponse,
  DiffResultParser,
  DiffSuccessResponse
} from './diffResultParser';
