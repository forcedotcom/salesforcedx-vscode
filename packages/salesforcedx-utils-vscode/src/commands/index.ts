/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

export { ChannelService } from './channelService';

import { NotificationService } from './notificationService';
export const notificationService = NotificationService.getInstance();
export { NotificationService };
export { ProgressNotification } from './progressNotification';
