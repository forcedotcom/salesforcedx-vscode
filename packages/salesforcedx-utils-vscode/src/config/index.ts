/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

/**
 * This file is intentionally left blank. ConfigSource and ConfigUtil are exported
 * in the root of salesforcedx-utils-vscode.
 */
