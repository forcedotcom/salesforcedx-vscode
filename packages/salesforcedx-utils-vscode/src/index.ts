/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

export * from './commands';
export {
  CompositeParametersGatherer,
  EmptyParametersGatherer
} from './commands/parameterGatherers';
export { EmptyPostChecker } from './commands/postconditionCheckers';
export {
  EmptyPreChecker,
  SfdxWorkspaceChecker
} from './commands/preconditionCheckers';
export {
  SfdxCommandletExecutor,
  LibraryCommandletExecutor
} from './commands/commandletExecutors';
export { SfdxCommandlet } from './commands/sfdxCommandlet';
export { ConfigSource, ConfigUtil } from './config/configUtil';
export {
  SETTING_CLEAR_OUTPUT_TAB,
  SFDX_CORE_CONFIGURATION_NAME,
  SFDX_CORE_EXTENSION_NAME
} from './constants';
export {
  OrgUserInfo,
  WorkspaceContextUtil
} from './context/workspaceContextUtil';
export {
  TelemetryService,
  TelemetryBuilder,
  TelemetryData,
  Properties,
  Measurements
} from './telemetry/telemetry';
export * from './helpers';
export { TraceFlags } from './helpers/traceFlags';
export { TelemetryReporter } from './telemetry/telemetryReporter';
export {
  hasRootWorkspace,
  getRootWorkspace,
  getRootWorkspacePath,
  workspaceUtils
} from './workspaces';
export * from './cli';
export * from './cli/commandExecutor';
export * from './i18n';
export * from './types';
export * from './date';
export * from './output';
export * from './predicates';
export * from './providers';
export * from './services';
