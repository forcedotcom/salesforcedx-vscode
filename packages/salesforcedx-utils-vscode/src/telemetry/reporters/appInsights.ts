/* eslint-disable header/header */
/*---------------------------------------------------------
 * Copyright (C) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------*/

'use strict';

import * as appInsights from 'applicationinsights';
import * as os from 'os';
import { Disposable, env, UIKind, version, workspace } from 'vscode';
import { WorkspaceContextUtil } from '../../context/workspaceContextUtil';
import { TelemetryReporter } from '../interfaces';

export class AppInsights extends Disposable implements TelemetryReporter {
  private appInsightsClient: appInsights.TelemetryClient | undefined;
  private userOptIn: boolean = false;
  private toDispose: Disposable[] = [];
  private uniqueUserMetrics: boolean = false;

  private static TELEMETRY_CONFIG_ID = 'telemetry';
  private static TELEMETRY_CONFIG_ENABLED_ID = 'enableTelemetry';

  constructor(
    private extensionId: string,
    private extensionVersion: string,
    key: string,
    enableUniqueMetrics?: boolean
  ) {
    super(() => this.toDispose.forEach(d => d && d.dispose()));
    if (enableUniqueMetrics) {
      this.uniqueUserMetrics = true;
    }
    this.updateUserOptIn(key);
    this.toDispose.push(
      workspace.onDidChangeConfiguration(() => this.updateUserOptIn(key))
    );
  }

  private updateUserOptIn(key: string): void {
    const config = workspace.getConfiguration(AppInsights.TELEMETRY_CONFIG_ID);
    if (
      this.userOptIn !==
      config.get<boolean>(AppInsights.TELEMETRY_CONFIG_ENABLED_ID, true)
    ) {
      this.userOptIn = config.get<boolean>(
        AppInsights.TELEMETRY_CONFIG_ENABLED_ID,
        true
      );
      if (this.userOptIn) {
        this.createAppInsightsClient(key);
      } else {
        // tslint:disable-next-line:no-floating-promises
        this.dispose();
      }
    }
  }

  private createAppInsightsClient(key: string) {
    // check if another instance is already initialized
    if (appInsights.defaultClient) {
      this.appInsightsClient = new appInsights.TelemetryClient(key);
      // no other way to enable offline mode
      this.appInsightsClient.channel.setUseDiskRetryCaching(true);
    } else {
      appInsights
        .setup(key)
        .setAutoCollectRequests(false)
        .setAutoCollectPerformance(false)
        .setAutoCollectExceptions(false)
        .setAutoCollectDependencies(false)
        .setAutoDependencyCorrelation(false)
        .setAutoCollectConsole(false)
        .setUseDiskRetryCaching(true)
        .start();
      this.appInsightsClient = appInsights.defaultClient;
    }

    this.appInsightsClient.commonProperties = this.getCommonProperties();
    if (this.uniqueUserMetrics && env) {
      this.appInsightsClient.context.tags['ai.user.id'] = env.machineId;
      this.appInsightsClient.context.tags['ai.session.id'] = env.sessionId;
      this.appInsightsClient.context.tags['ai.cloud.roleInstance'] =
        'DEPRECATED';
    }

    // check if it's an Asimov key to change the endpoint
    if (key && key.indexOf('AIF-') === 0) {
      this.appInsightsClient.config.endpointUrl =
        'https://vortex.data.microsoft.com/collect/v1';
    }
  }

  private getCommonProperties(): { [key: string]: string } {
    const commonProperties = Object.create(null);
    commonProperties['common.os'] = os.platform();
    commonProperties['common.platformversion'] = (os.release() || '').replace(
      /^(\d+)(\.\d+)?(\.\d+)?(.*)/,
      '$1$2$3'
    );
    const cpus = os.cpus();
    if (cpus && cpus.length > 0) {
      commonProperties[
        'common.cpus'
      ] = `${cpus[0].model}(${cpus.length} x ${cpus[0].speed})`;
    }
    commonProperties['common.systemmemory'] = `${(
      os.totalmem() /
      (1024 * 1024 * 1024)
    ).toFixed(2)} GB`;
    commonProperties['common.extname'] = this.extensionId;
    commonProperties['common.extversion'] = this.extensionVersion;
    if (env) {
      commonProperties['common.vscodemachineid'] = env.machineId;
      commonProperties['common.vscodesessionid'] = env.sessionId;
      commonProperties['common.vscodeversion'] = version;
      if (env.uiKind) {
        commonProperties['common.vscodeuikind'] = UIKind[env.uiKind];
      }
    }
    return commonProperties;
  }

  public sendTelemetryEvent(
    eventName: string,
    properties?: { [key: string]: string },
    measurements?: { [key: string]: number }
  ): void {
    if (this.userOptIn && eventName && this.appInsightsClient) {
      const orgId = WorkspaceContextUtil.getInstance().orgId;
      if (orgId && properties) {
        properties.orgId = orgId;
      } else if (orgId) {
        properties = { orgId };
      }
      this.appInsightsClient.trackEvent({
        name: `${this.extensionId}/${eventName}`,
        // tslint:disable-next-line:object-literal-shorthand
        properties,
        // tslint:disable-next-line:object-literal-shorthand
        measurements
      });
    }
  }

  public sendExceptionEvent(
    exceptionName: string,
    exceptionMessage: string,
    measurements?: { [key: string]: number }
  ): void {
    if (this.userOptIn && exceptionMessage && this.appInsightsClient) {
      const error = new Error();
      error.name = `${this.extensionId}/${exceptionName}`;
      error.message = exceptionMessage;
      error.stack = 'DEPRECATED';

      const orgId = WorkspaceContextUtil.getInstance().orgId || '';
      const properties = { orgId };
      this.appInsightsClient.trackException({
        exception: error,
        properties,
        measurements
      });
    }
  }

  public dispose(): Promise<any> {
    const flushEventsToAI = new Promise<any>(resolve => {
      if (this.appInsightsClient) {
        this.appInsightsClient.flush({
          callback: () => {
            // all data flushed
            this.appInsightsClient = undefined;
            resolve(void 0);
          }
        });
      } else {
        resolve(void 0);
      }
    });
    return flushEventsToAI;
  }
}
