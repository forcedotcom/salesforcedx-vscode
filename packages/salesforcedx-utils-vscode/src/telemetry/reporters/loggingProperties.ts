/*
 * Copyright (c) 2024, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

export type CommonProperties = {
  'common.os': string;
  'common.platformversion': string;
  'common.cpus'?: string;
  'common.systemmemory': string;
  'common.extname': string;
  'common.extversion': string;
  'common.vscodemachineid'?: string;
  'common.vscodesessionid'?: string;
  'common.vscodeversion'?: string;
  'common.vscodeuikind'?: string;
};

export type InternalProperties = {
  'sfInternal.hostname': string;
  'sfInternal.username': string;
};
