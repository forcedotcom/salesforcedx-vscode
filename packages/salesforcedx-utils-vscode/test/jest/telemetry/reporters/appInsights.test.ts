/*
 * Copyright (c) 2024, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
import * as os from 'os';
import { workspace } from 'vscode';
import { WorkspaceContextUtil } from '../../../../src';
import { AppInsights } from '../../../../src/telemetry/reporters/appInsights';
import { CommonProperties, InternalProperties } from '../../../../src/telemetry/reporters/loggingProperties';

describe('AppInsights', () => {
  const fakeExtensionId = 'anExtensionId';
  const fakeExtensionVersion = '0.10.0';
  const fakeUserId = '45gkjnbxsbchdnv34sbcishsm';
  const fakeKey = 'testKey';

  describe('sendTelemetryEvent and sendExceptionEvent', () => {
    let getInstanceMock: jest.SpyInstance;
    const dummyOrgId = '000dummyOrgId';
    const getMock = jest.fn().mockReturnValueOnce(true);
    const fakeConfig: any = { get: getMock };

    let appInsights: AppInsights;
    const trackExceptionMock = jest.fn();
    const trackEventMock = jest.fn();

    beforeEach(() => {
      // Arrange
      getInstanceMock = jest
        .spyOn(WorkspaceContextUtil, 'getInstance')
        .mockReturnValue({
          orgId: dummyOrgId
        } as any);

      jest.spyOn(workspace, 'getConfiguration').mockReturnValue(fakeConfig);
      jest
        .spyOn(AppInsights.prototype as any, 'updateUserOptIn')
        .mockReturnValue('');
    });

    it('should send telemetry data to appInsightsClient.trackEvent', () => {
      appInsights = new AppInsights(fakeExtensionId, fakeExtensionVersion, '', fakeUserId, false);
      (appInsights as any).userOptIn = true;
      (appInsights as any).appInsightsClient = {
        trackException: trackExceptionMock,
        trackEvent: trackEventMock
      };

      // Act
      appInsights.sendTelemetryEvent('Dummy Telemetry Event', {}, {});

      // Assert
      expect(getInstanceMock).toHaveBeenCalledTimes(1);
      expect(trackEventMock).toHaveBeenCalledTimes(1);
      expect(trackEventMock.mock.calls[0][0]).toMatchSnapshot();
    });

    it('should send orgId to appInsightsClient.trackException', () => {
      // Act
      appInsights.sendExceptionEvent(
        'Dummy Exception',
        'a dummy exception occurred'
      );

      // Assert
      expect(getInstanceMock).toHaveBeenCalledTimes(1);
      expect(trackExceptionMock).toHaveBeenCalledTimes(1);
      expect(trackExceptionMock.mock.calls[0][0]).toMatchSnapshot();
    });
  });

  describe('dispose', () => {
    let appInsights: AppInsights;
    const flushMock = jest.fn();
    const appInsightsClientMock = {
      flush: flushMock
    };

    beforeEach(() => {
      appInsights = new AppInsights(
        fakeExtensionId,
        fakeExtensionVersion,
        'aKey',
        fakeUserId,
        false
      );
      (appInsights as any).appInsightsClient = appInsightsClientMock;
    });

    it('should flush events to appInsightsClient and resolve', () => {
      const expectedPromiseResult = Promise.resolve();

      const disposePromise = appInsights.dispose();

      expect(flushMock).toHaveBeenCalledTimes(1);
      expect(disposePromise).toEqual(expectedPromiseResult);
    });

    it('should resolve immediately if appInsightsClient is undefined', () => {
      (appInsights as any).appInsightsClient = undefined;
      const expectedPromiseResult = Promise.resolve();

      const disposePromise = appInsights.dispose();

      expect(disposePromise).toEqual(expectedPromiseResult);
    });
  });

  describe('AppInsights - getCommonProperties', () => {
    let appInsights: AppInsights;
    let commonProperties: CommonProperties;

    beforeEach(() => {
      appInsights = new AppInsights(fakeExtensionId, fakeExtensionVersion, fakeKey, fakeUserId, false);
      commonProperties = appInsights['getCommonProperties']();
    });

    it('should return common system properties', () => {
      expect(typeof commonProperties).toBe('object');
    });

    it('should return extname that was passed in params', () => {
      expect(commonProperties['common.extname']).toBe(fakeExtensionId);
    });

    it('should return extVersion passed in params', () => {
      expect(commonProperties['common.extversion']).toBe(fakeExtensionVersion);
    });
  });

  describe('AppInsights - getInternalProperties', () => {
    let appInsights: AppInsights;
    let internalProperties: InternalProperties;

    beforeEach(() => {
      jest.spyOn(os, 'hostname').mockReturnValue('test.internal.salesforce.com');
      jest.spyOn(os, 'userInfo').mockReturnValue({
        username: 'testuser',
        uid: 1001,
        gid: 1001,
        shell: '/bin/bash',
        homedir: '/home/testuser'
      });
      appInsights = new AppInsights(fakeExtensionId, fakeExtensionVersion, fakeKey, fakeUserId, false);
      internalProperties = appInsights['getInternalProperties']();
    });

    afterEach(() => {
      jest.clearAllMocks();
    });

    it('should return internal properties', () => {
      expect(typeof internalProperties).toBe('object');
    });

    it('should return hostname', () => {
      expect(internalProperties['sfInternal.hostname']).toBe('test.internal.salesforce.com');
    });

    it('should return username', () => {
      expect(internalProperties['sfInternal.username']).toBe('testuser');
    });
  });

  describe('AppInsights - aggregateLoggingProperties', () => {
    let appInsights: AppInsights;

    beforeEach(() => {
      jest
        .spyOn(os, 'hostname')
        .mockReturnValue('test.internal.salesforce.com');
      jest.spyOn(os, 'cpus').mockReturnValue([
        {
          model: 'AMD EPYC 7763 64-Core Processor',
          speed: 3242,
          times: {
            user: 100000,
            nice: 0,
            sys: 100000,
            idle: 1000000,
            irq: 0
          }
        }
      ]);
      jest.spyOn(os, 'userInfo').mockReturnValue({
        username: 'testuser',
        uid: 1001,
        gid: 1001,
        shell: '/bin/bash',
        homedir: '/home/testuser'
      });
      appInsights = new AppInsights(fakeExtensionId, fakeExtensionVersion, fakeKey, fakeUserId, false);
    });

    afterEach(() => {
      jest.clearAllMocks();
    });

    it('should return properties', () => {
      expect(typeof appInsights['aggregateLoggingProperties']()).toBe('object');
    });

    it('should return common and internal properties when is internal user', () => {
      const commonProps = appInsights['getCommonProperties']();
      const internalProps = appInsights['getInternalProperties']();
      const result = appInsights['aggregateLoggingProperties']();
      expect(result).toEqual({...commonProps, ...internalProps});
    });

    it('should return common properties when is not internal user', () => {
      jest.spyOn(os, 'hostname').mockReturnValue('test.salesforce.com');
      const commonProps = appInsights['getCommonProperties']();
      const result = appInsights['aggregateLoggingProperties']();
      expect(result).toEqual(commonProps);
    });
  });
});
