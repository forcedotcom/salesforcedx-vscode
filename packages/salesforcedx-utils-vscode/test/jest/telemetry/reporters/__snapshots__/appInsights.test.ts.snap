// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppInsights sendTelemetryEvent and sendExceptionEvent should send orgId to appInsightsClient.trackException 1`] = `
{
  "exception": [anExtensionId/Dummy Exception: a dummy exception occurred],
  "measurements": undefined,
  "properties": {
    "orgId": "000dummyOrgId",
  },
}
`;

exports[`AppInsights sendTelemetryEvent and sendExceptionEvent should send telemetry data to appInsightsClient.trackEvent 1`] = `
{
  "measurements": {},
  "name": "anExtensionId/Dummy Telemetry Event",
  "properties": {
    "orgId": "000dummyOrgId",
  },
}
`;
